snippetpixie: init at 1.2.2

ianmjones opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Add Snippet Pixie as a Nix package.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notes

* Requires `services.gnome3.at-spi2-core.enable = true;` on NixOS if not already up and running in WM/DE.
* I'm the author of Snippet Pixie.
* My first PR for anything Nix/NixPkgs/NixOS! :tada:

ianmjones committed:
snippetpixie: init at 1.2.2

ianmjones commented:
@worldofpeace I'm not clear on the etiquette, hope you don't mind me pinging you here, but just thought this PR might be one you'd be best equipped to review if you get a chance as I see you've worked on a lot of pantheon/elementary PRs?

Apologies if this note is bad form, couldn't find any documentation on how to find a good reviewer for a new package.

worldofpeace commented:
> @worldofpeace I'm not clear on the etiquette, hope you don't mind me pinging you here, but just thought this PR might be one you'd be best equipped to review if you get a chance as I see you've worked on a lot of pantheon/elementary PRs?
> 
> Apologies if this note is bad form, couldn't find any documentation on how to find a good reviewer for a new package.

Oh it's not in bad form in all. Actually, you've found the perfect person because I maintain pantheon in nixpkgs, and also can co-maintain any elementary appcenter application.
Ref https://github.com/NixOS/nixpkgs/blob/master/pkgs/desktops/pantheon/default.nix#L59.

In the future I hope we add maintainer teams
* https://github.com/NixOS/nixpkgs/pull/72125

so people can discover who to request reviews on a certain domain.

worldofpeace commented on maintainers/maintainer-list.nix:
```diff
@@ -2862,6 +2862,12 @@
+    github = "ianmjones";
+    githubId = 4710;
+    name = "Ian M. Jones";
+  };
```
Could you add yourself to the mainatiner's list in a separate commit like?
```
maintainers: add ianmjones
```

ianmjones replied:
As in, just remove in one commit and then re-apply in separate commit to enable cherry picking?

worldofpeace replied:
like the git log is
```
maintainers: add ianmjones
snippetpixie: init at 1.2.2
```

First commit touches `maintainer-list.nix` only.

worldofpeace commented on pkgs/tools/text/snippetpixie/default.nix:
```diff
@@ -0,0 +1,81 @@
+  ibus,
+  json-glib,
+  pantheon
+  }:
```
```suggestion
}:
```

worldofpeace commented on pkgs/tools/text/snippetpixie/default.nix:
```diff
@@ -0,0 +1,81 @@
+  dbus,
+  ibus,
+  json-glib,
+  pantheon
```
your comma's are backwards here for long attribute sets.

> https://nixos.org/nixpkgs/manual/#chap-conventions
>  but if they don't fit on one line they're written as: 

Example is out of date but it's like in haskell
```
{ stdenv
, something
, something
, something
, ... 
```

worldofpeace commented:
It also seems you've by in large authored this
https://github.com/bytepixie/snippetpixie/graphs/contributors

That's pretty cool, thanks for submitting it to nixpkgs :+1:

worldofpeace commented on pkgs/tools/text/snippetpixie/default.nix:
```diff
@@ -0,0 +1,81 @@
+    homepage    = https://www.snippetpixie.com;
+    license     = licenses.gpl2Plus;
+    maintainers = with maintainers; [ ianmjones ];
+    platforms   = platforms.linux;
```
Can we not align these vertically?
I believe https://nixos.org/nixpkgs/manual/#chap-conventions feigns this, but it's largely discouraged.

worldofpeace commented on pkgs/tools/text/snippetpixie/default.nix:
```diff
@@ -0,0 +1,81 @@
+    vala
+    pkgconfig
+    wrapGAppsHook
+    cmake
```
why do we need cmake? I know meson will use it if it can't find a dep through pkgconfig, but we have that too.

worldofpeace commented on pkgs/tools/text/snippetpixie/default.nix:
```diff
@@ -0,0 +1,81 @@
+    pantheon.granite
+    pantheon.elementary-gtk-theme
+    pantheon.elementary-icon-theme
+  ];
```
can you add `glib`? It's in the dependencies https://github.com/bytepixie/snippetpixie/blob/9f475d327ae5d3073ca16a2e34769f9cefc12792/meson.build#L30, and it's namespace is used in at least one call.

worldofpeace commented on pkgs/tools/text/snippetpixie/default.nix:
```diff
@@ -0,0 +1,81 @@
+    '';
+    homepage    = https://www.snippetpixie.com;
+    license     = licenses.gpl2Plus;
+    maintainers = with maintainers; [ ianmjones ];
```
```suggestion
    maintainers = with maintainers; [ 
      ianmjones 
    ] ++ pantheon.maintainers;
```

worldofpeace reviewed (changes requested):
Everything seems good, left some comments on style and few adjustments.

I ran this locally and I did notice a warning immediately
```
(com.github.bytepixie.snippetpixie:18392): dbind-WARNING **: 05:47:48.982: AT-SPI: Unable to open bus connection: Failed to connect to socket /run/user/1000/at-spi2-RGR4C0/socket: No such file or directory

(com.github.bytepixie.snippetpixie:18392): dbind-WARNING **: 05:47:48.985: AT-SPI: Unable to open bus connection: Failed to connect to socket /run/user/1000/at-spi2-HDX3C0/socket: No such file or directory

(com.github.bytepixie.snippetpixie:18392): dbind-WARNING **: 05:47:48.985: AT-SPI: Unable to open bus connection: Failed to connect to socket /run/user/1000/at-spi2-UGB4C0/socket: No such file or directory

(com.github.bytepixie.snippetpixie:18392): dbind-WARNING **: 05:47:48.985: AT-SPI: Unable to open bus connection: Failed to connect to socket /run/user/1000/at-spi2-DL03C0/socket: No such file or directory

(com.github.bytepixie.snippetpixie:18392): dbind-WARNING **: 05:47:48.986: AT-SPI: Unable to open bus connection: Failed to connect to socket /run/user/1000/at-spi2-HCZ3C0/socket: No such file or directory

(com.github.bytepixie.snippetpixie:18392): dbind-WARNING **: 05:47:48.986: AT-SPI: Unable to open bus connection: Failed to connect to socket /run/user/1000/at-spi2-V033C0/socket: No such file or directory

(com.github.bytepixie.snippetpixie:18392): dbind-WARNING **: 05:47:48.986: AT-SPI: Unable to open bus connection: Failed to connect to socket /run/user/1000/at-spi2-WIZ3C0/socket: No such file or directory

(com.github.bytepixie.snippetpixie:18392): dbind-WARNING **: 05:47:48.987: AT-SPI: Unable to open bus connection: Failed to connect to socket /run/user/1000/at-spi2-DE53C0/socket: No such file or directory
```

and this does appear to have scattered several `at-spi2-*` temporary directories.

I do have `at-spi-bus-launcher` running.

Another warning was
```
** (com.github.bytepixie.snippetpixie:19326): WARNING **: 05:53:30.036: Application.vala:821: Could not find desktop file with name: com.github.bytepixie.snippetpixie.desktop
```
This is weird because [`DesktopAppInfo`](https://github.com/bytepixie/snippetpixie/blob/c5a2839252dd76d626a7392e706f79b9b0a8a2ea/src/Application.vala#L818) should work  because the docs state
> https://valadoc.org/gio-unix-2.0/GLib.DesktopAppInfo.DesktopAppInfo.html
>  GIO is looking for a desktop file with this name in the `applications` subdirectories of the XDG data directories (i.e. the directories specified in the `XDG_DATA_HOME` and `XDG_DATA_DIRS` environment variables). 

and to support running applications uninstalled, for purity, our wrapper scripts set
```
export XDG_DATA_DIRS='/nix/store/fyajylfficz2kd1lnbbrg4yk4xyl2lw2-snippetpixie-1.2.2/share'${XDG_DATA_DIRS:+':'}$XDG_DATA_DIRS
```

ianmjones committed:
maintainers: add ianmjones

ianmjones commented:
Thanks for the review @worldofpeace, hopefully I've addressed most of your feedback.

> and this does appear to have scattered several at-spi2-* temporary directories.
> 
> I do have at-spi-bus-launcher running.

Hmm, I don't see all the warnings you see, but I suspect I'm running the build differently...

> Another warning was
> 
> `** (com.github.bytepixie.snippetpixie:19326): WARNING **: 05:53:30.036: Application.vala:821: Could not find desktop file with name: com.github.bytepixie.snippetpixie.desktop`
> 
> This is weird because DesktopAppInfo should work because the docs state...[snip]

I don't get this warning either, the desktop file is created dynamically if missing, for me that's in ~/.config/autostart/.

So I think that's because I've just tested it as per the quick start guide after a `nix-env -f . -iA snippetpixie`, which means it's a local nix profile install rather than read-only system level install?

How do you build/install/test so that I can try and reproduce?

worldofpeace commented:
@ianmjones You can run it like
```
nix-build -A snippetpixie
./result/bin/$binary
$binary
```

and I think it should be reproduced in nix-shell too
```
nix-build -A snippetpixie
nix-shell -p ./result
$binary
```

Oh, I didn't emphasize enough that it's run uninstalled. That's why I mentioned that application is wrapped to have the nix store path with its XDG_DATA_DIR.

worldofpeace commented:
@ianmjones Running things like this is kinda a feature in nixos we provide (purity), so in some situations it doesn't work correctly. Some issues require a slight change in the sourcecode, however I didn't notice anything up.

worldofpeace commented:
I tested it installed and the
```
** (com.github.bytepixie.snippetpixie:4421): WARNING **: 18:08:44.607: Application.vala:821: Could not find desktop file with name: com.github.bytepixie.snippetpixie.desktop
```

error goes away. Though for the reason at
> 

> This is weird because [`DesktopAppInfo`](https://github.com/bytepixie/snippetpixie/blob/c5a2839252dd76d626a7392e706f79b9b0a8a2ea/src/Application.vala#L818) should work because the docs state
> 
> > https://valadoc.org/gio-unix-2.0/GLib.DesktopAppInfo.DesktopAppInfo.html
> > GIO is looking for a desktop file with this name in the `applications` subdirectories of the XDG data directories (i.e. the directories specified in the `XDG_DATA_HOME` and `XDG_DATA_DIRS` environment variables).
> 
> and to support running applications uninstalled, for purity, our wrapper scripts set
> 
> ```
> export XDG_DATA_DIRS='/nix/store/fyajylfficz2kd1lnbbrg
> ```

the child process (the actual program) will inherit adding the `share` dir of the package to XDG_DATA_DIRS via the wrapGAppsHook script. Meaning, for reasons documented in glib, this shouldn't be an issue.

ianmjones commented:
@worldofpeace Thanks for the steps to reproduce the error, I can reproduce, but I also have no idea why it doesn't find the default desktop file that's right there in `./result/share/applications/com.github.bytepixie.snippetpixie.desktop` when running `./result/bin/com.github.bytepixie.snippetpixie`.

When running the installed version all seems to be fine, I get a properly generated `~/.config/autostart/com.github.bytepixie.snippetpixie.desktop` file based on the original. ¯\\_(ツ)_/¯

For example, I got that shrug above with my ``shrug` `` abbreviation while writing this in Quilter, running in SwayWM, on NixOS 19.09!

worldofpeace commented:
> @worldofpeace Thanks for the steps to reproduce the error, I can reproduce, but I also have no idea why it doesn't find the default desktop file that's right there in `./result/share/applications/com.github.bytepixie.snippetpixie.desktop` when running `./result/bin/com.github.bytepixie.snippetpixie`.
> 
> When running the installed version all seems to be fine, I get a properly generated `~/.config/autostart/com.github.bytepixie.snippetpixie.desktop` file based on the original. ¯\_(ツ)_/¯
> 
> For example, I got that shrug above with my `` shrug` `` abbreviation while writing this in Quilter, running in SwayWM, on NixOS 19.09!

Yeah, it has me confused also, should just work.

I think we can add the issue to your tracker and merge, since it does function as expected installed.
cc @jtojnar, am I confusing how works https://valadoc.org/gio-unix-2.0/GLib.DesktopAppInfo.DesktopAppInfo.html?

@ianmjones Could you rewrite your history to https://github.com/NixOS/nixpkgs/pull/75472#discussion_r356797914?

jtojnar commented:
@worldofpeace If I recall correctly from my experiments with the interface, it also does some validation. Most notably whether the command referenced in `Exec` is in `PATH`.

worldofpeace commented:
Found the code in question for glib
* https://gitlab.gnome.org/GNOME/glib/blob/master/gio/gdesktopappinfo.c#L1785

Guess this will be an issue for glib and nixos, so I guess that's out of scope per this PR.

ianmjones commented:
> @ianmjones Could you rewrite your history to [#75472 (comment)](https://github.com/NixOS/nixpkgs/pull/75472#discussion_r356797914)?

@worldofpeace Had a go, first time I've ever done that, so hopefully it's ok? I applied the maintainer list change first in the pick list, but it's date puts it second, hopefully that's not an issue.

worldofpeace commented:
> > @ianmjones Could you rewrite your history to [#75472 (comment)](https://github.com/NixOS/nixpkgs/pull/75472#discussion_r356797914)?
> 
> @worldofpeace Had a go, first time I've ever done that, so hopefully it's ok? I applied the maintainer list change first in the pick list, but it's date puts it second, hopefully that's not an issue.

Yeah, that's just github being silly. It as actually ordered before the `init`.
